### PR TITLE
fix(49223): checker.getTypeAtLocation for ExpressionWithTypeArguments returns an error any type

### LINKED ISSUE #49223: `checker.getTypeAtLocation` for `ExpressionWithTypeArguments` returns an error any type
# Bug Report

### 🔎 Search Terms

`ExpressionWithTypeArguments`, `getTypeAtLocation`, `checker`, api, error any.

### 🕗 Version & Regression Information

- This is a bug in the `@next` version of TS.

### ⏯ Playground Link

Because this is a bug with the API specifically, here is a link to the ast explorer.
https://ts-ast-viewer.com/#code/GYVwdgxgLglg9mABAWwIYGsCmAhOAPAHgBUA+ACgDdUAbETALkSIEpEBvAKEUQCdMoQPJG0RVamRAF8A3B0kcOEBAGcoKDJgDKUHjDABzXHkQBedViMFVugyWlA

image

### 💻 Code

```ts
function makeBox<T>(value: T) {
  return { value };
}

const makeStringBox = makeBox<string>;


const sett = Set<string>;
```

### 🙁 Actual behavior

`checker.getTypeAtLocation` for the `ExpressionWithTypeArguments` node returns an error `any`.

### 🙂 Expected behavior

`checker.getTypeAtLocation` for the `ExpressionWithTypeArguments` node returns the relevant type of the expression, which would be `(value: string) => {value: string}`

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #49223

### CODE REVIEW

**Comment by bradennapier** [comment]
That was quick! Thanks!

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -2016,6 +2016,8 @@ namespace ts {
             case SyntaxKind.AwaitExpression:
             case SyntaxKind.MetaProperty:
                 return true;
+            case SyntaxKind.ExpressionWithTypeArguments:
+                return !(isPartOfTypeNode(node.parent) || isHeritageClause(node.parent));
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts:2020]: Why is it not an expression in a heritage clause? Feels like it must be the case that it's an expression in an `extends` clause, right? @ahejlsberg

Comment (ahejlsberg) [reply src/compiler/utilities.ts:2020]: The `isPartOfTypeNode(node.parent)` implies that we're in a heritage clause (of an interface), so it's redundant here. In other words, the code is equivalent to just `return !isHeritageClause(node.parent)`. That said, it's probably more correct to change the code to `return !isPartOfTypeNode(node.parent)`, which would say that if it's not a type, it must be an expression. That would additionally return true when the expression with type arguments occurs in a class heritage clause.

Comment (DanielRosenwasser) [reply src/compiler/utilities.ts:2020]: So just

```suggestion
                return !isPartOfTypeNode(node.parent);
```

?

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.7

**Comment by DanielRosenwasser** [comment]
Ugh, sorry. Would you be able to accept baselines?

File: src/compiler/utilities.ts (hunk 1 above)
Comment (DanielRosenwasser) [reply src/compiler/utilities.ts]: @andrewbranch commented that that's likely desirable for classes, but not necessarily for types.

Comment (andrewbranch) [reply src/compiler/utilities.ts]: I used [this example](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEA3W8BCUDOIBc8De8AdiAO4A8AKgHwAUAlLlIQJ7wC+AUAC7MAOC6LPAC88AEQALKPCnwMEgPYBXCMHhcJASwzwARgg0G+CKF3gADEAA8uIQsB2CQ5gPxiA3Bw6RMOgMLw1rb2jpggZBhcMJqEAOZU+F6cQA) to demonstrate why using `getTypeFromTypeNode` would be wrong there. If something is an expression, you have to get the type of the expression.

**Comment by DanielRosenwasser** [comment]
Doing this pretty much wrecks every `.types` baseline so that something like

```ts
class ContextMenu extends EventManager {
  // ...
}
```

has the following diff

```diff
  >ContextMenu : ContextMenu
- >EventManager : EventManager
+ >EventManager : typeof EventManager
+ >EventManager : typeof EventManager
```

**Comment by DanielRosenwasser** [comment]
It also breaks how find-all-refs seems to work.

```diff
  // === /tests/cases/fourslash/findAllRefsForStaticInstanceMethodInheritance.ts ===
  // class X{
  // 	/*FIND ALL REFS*/[|foo|](): void{}
  // }
  // 
  // class Y extends X{
  // 	static foo(): void{}
  // }
  // 
  // class Z extends Y{
  // 	static foo(): void{}
- // 	[|foo|](): void{}
+ // 	foo(): void{}
  // }
  // 
  // const x = new X();
  // const y = new Y();
  // const z = new Z();
  // x.[|foo|]();
  // y.[|foo|]();
- // z.[|foo|]();
+ // z.foo();
  // Y.foo();
  // Z.foo();
```

@ahejlsberg @RyanCavanaugh to be honest, I am considering splitting this into a 4.7.4 change unless others can help guide this PR today.

**Comment by DanielRosenwasser** [comment]
Unless we just revert my change at https://github.com/microsoft/TypeScript/pull/49284/commits/1394d807713a94bb08fef5b867d05cef6ee5cdf7 and go with @a-tarasyuk's original change.

**Comment by ahejlsberg** [comment]
I would go with `return !isHeritageClause(node.parent);` instead of the original change (since the other part of the expression is redundant).

**Comment by DanielRosenwasser** [comment]
Everything seems to pass with just that.

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.7

**Review by DanielRosenwasser — APPROVED** [review]

**Pull request merged**